llms/mlx_lm/server.py

```diff
@@ -0,0 +1,319 @@
+                        "logprobs": None,
+                        "finish_reason": None,
+                    }
+                ],
+                "usage": {
+                    "prompt_tokens": len(prompt),
+                    "completion_tokens": len(tokens),
+                    "total_tokens": len(prompt) + len(tokens),
+                },
+            }
+
+            return response
+        else:
+            self.send_response(200)
+            self.send_header("Content-type", "text/event-stream")
+            self.send_header("Cache-Control", "no-cache")
+            self.end_headers()
+            accumulated_tokens = []
+            current_generated_text_index = 0
+            for token in generate(
+                prompt,
+                _model,
+                temperature,
+                stop_id_sequences,
+                eos_token_id,
+                max_tokens,
+                top_p=top_p,
+            ):
+                # This is a workaround because the llama tokenizer omitted spaces during decoding token by token.
+                accumulated_tokens.append(token)
+                generated_text = _tokenizer.decode(accumulated_tokens)
+                next_chunk = generated_text[current_generated_text_index:]
+                current_generated_text_index = len(generated_text)
+                response = {
+                    "id": chat_id,
+                    "object": "chat.completion.chunk",
+                    "created": int(time.time()),
+                    "model": requested_model,
+                    "system_fingerprint": f"fp_{uuid.uuid4()}",
+                    "choices": [
+                        {
+                            "index": 0,
+                            "delta": {"role": "assistant", "content": next_chunk},
+                            "logprobs": None,
+                            "finish_reason": None,
+                        }
+                    ],
+                }
+                try:
+                    self.wfile.write(f"data: {json.dumps(response)}\n\n".encode())
+                    self.wfile.flush()
+                except Exception as e:
+                    print(e)
+                    break
+
+            self.wfile.write(f"data: [DONE]\n\n".encode())
+            self.wfile.flush()
+
+
+def run(host: str, port: int, server_class=HTTPServer, handler_class=APIHandler):
+    server_address = (host, port)
+    httpd = server_class(server_address, handler_class)
+    print(f"Starting httpd at {host} on port {port}...")
+    httpd.serve_forever()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="MLX Http Server.")
+    parser.add_argument(
+        "--model",
+        type=str,
+        required=True,
+        help="The path to the MLX model weights, tokenizer, and config",
+    )
+    parser.add_argument(
+        "--adapter-file",
+        type=str,
+        help="Optional path for the trained adapter weights.",
```

Comment:
Doesn't it need a default to be optional?

Reply:
As long as it is not required, it will default to None. If you prefer explicit, I can add a default value of None.

Reply:
No it's fine the way you have it.. I did not know that 😄